packages/context/test/unit/context.ts

```diff
@@ -63,7 +64,100 @@ describe('Context', () => {
+        ctx.bind('foo').toProvider(MockValueProvider);
+        expect(ctx.contains('foo')).to.be.true();
+      });
+
+      it('binding.getValue() injects constructor values in the Provider Class', async () => {
+        ctx.bind('msg').to('hello');
+        ctx.bind('foo').toProvider(MockValueProvider);
+        const value: string = await ctx.getBinding('foo').getValue(ctx);
+        expect(value).to.equal('hello world');
+      });
+
+      it('binding.getValue() returns the same value as provider.value()', async () => {
+        ctx.bind('msg').to('hello');
+        ctx.bind('foo').toProvider(MockValueProvider);
+        const providerInstance = new MockValueProvider('hello');
+        const value: string = await ctx.getBinding('foo').getValue(ctx);
+        expect(value).to.equal(providerInstance.value(ctx));
+      });
+    });
+
+    describe('FunctionProvider', () => {
+      let provider: MockFunctionProvider;
+      let date: Date;
+      let prefix: string;
+
+      beforeEach(givenProvider);
+
+      it('binds a provider class to a context', () => {
+        ctx.bind('foo').toProvider(MockFunctionProvider);
+        expect(ctx.contains('foo')).to.be.true();
+      });
+
+      it('binding.getValue() injects constructor values in the Provider Class', async () => {
+        ctx.bind('now').to(date);
+        ctx.bind('prefix').to(prefix);
+        ctx.bind('foo').toProvider(MockFunctionProvider);
+        const value: string = await ctx.getBinding('foo').getValue(ctx);
+        expect(value).to.equal(prefix + ' ' + date);
+      });
+
+      it('binding.getValue() returns the same value as provider.value()', async () => {
+        ctx.bind('now').to(date);
+        ctx.bind('prefix').to(prefix);
+        ctx.bind('foo').toProvider(MockFunctionProvider);
+        const value: string = await ctx.getBinding('foo').getValue(ctx);
+        const providerInstance = new MockFunctionProvider(date, prefix);
+        const providerInstanceValue = await providerInstance.value(ctx);
+        expect(value).to.equal(providerInstanceValue);
+      });
+
+      function givenProvider() {
+        date = new Date();
+        prefix = 'Time is';
+        provider = new MockFunctionProvider(date, prefix);
+      }
+    });
+
+    describe('ConstructorProvider', () => {
+      let provider: MockConstructorProvider;
+
+      beforeEach(givenProvider);
+
+      it('binds a provider class to a context', () => {
+        ctx.bind('foo').toProvider(MockConstructorProvider);
+        expect(ctx.contains('foo')).to.be.true();
+      });
+
+      it('binding.getValue() injects constructor values in the Provider Class', async () => {
+        ctx.bind('arg1').to('hello');
+        ctx.bind('param').to('world');
+        ctx.bind('foo').toProvider(MockConstructorProvider);
+        const value: string = await ctx.getBinding('foo').getValue(ctx);
+        expect(value).to.be.instanceof(MockLoopbackController);
```

Comment:
Value has been declared with type `string`. It does not make sense to me to expect a string to be an instance of a controller class. Should `value` be typed as `Object` or even `MockLoopbackController`?